src/inmanta/ast/type.py

```diff
@@ -170,35 +263,45 @@ class Primitive(Type):
 
     def __init__(self) -> None:
         Type.__init__(self)
-        self.try_cast_functions: Sequence[Callable[[Optional[object]], object]] = []
+        self.cast_function: Callable[[Optional[object]], object] = cast_not_implemented
```

Comment:
I simplified this a bit

Reply:
We used to have more than one here, I hadn't realized that that was no longer the case. Nice! I like how it makes the `cast` method implementation so much simpler. As to the not implemented path, I wonder if we can simply go for an abstract property instead. Iirc you can override that with a simple attribute in the child classes, but it's been a while since I've tried.

Reply:
I would like something abc-like, but that ship,...

Reply:
Well, this could still become one, but I see your point. Ok, we leave this as is then.